test/test_jit_autocast.py

```diff
@@ -0,0 +1,255 @@
+                with autocast_off:
+                    e = torch.mm(c, d)
+            return torch.mm(e, e)
+        fn(self.a_fp32, self.b_fp32, self.c_fp32, self.d_fp32)
+
+    def test_conditional_autocast(self):
+        @torch.jit.script
+        def fn(a, b):
+            autocast_on = autocast(enabled=True)
+            autocast_off = autocast(enabled=False)
+            with autocast_on if a[0][0] > 0.5 else autocast_off:
+                return torch.mm(a, b)
+        # conditional autocast expressions are not supported
+        with self.assertRaises(RuntimeError):
+            fn(self.a_fp32, self.b_fp32)
+
+    def test_nested_autocast(self):
+        @torch.jit.script
+        def fn(a, b, c, d):
+            with autocast(enabled=False):
+                e = torch.mm(a, b)
+                with autocast(enabled=True):
+                    f = torch.mm(e, c)
+                    with autocast(enabled=False):
+                        g = torch.mm(e, d)
+            return e, f, g
+        e, f, g = fn(self.a_fp32, self.b_fp32, self.c_fp32, self.d_fp32)
+        self.assertEqual(e.dtype, torch.float32)
+        self.assertEqual(f.dtype, torch.float16)
+        self.assertEqual(g.dtype, torch.float32)
+
+    def test_implicitly_nested_autocast(self):
+        @torch.jit.script
+        def fn(a, b):
+            with autocast(enabled=False), autocast(enabled=True):
+                return torch.mm(a, b)
+        result = fn(self.a_fp32, self.b_fp32)
+        self.assertEqual(result.dtype, torch.float16)
+
+    def test_reused_autocast(self):
+        @torch.jit.script
+        def fn(a, b, c, d):
+            autocast_instance = autocast(enabled=True)
+            with autocast_instance:
+                e = torch.mm(a, b)
+                with autocast_instance:
+                    e = torch.mm(c, d)
+                    f = torch.mm(d, e)
+            g = torch.mm(e, f)
+            return e, f, g
+        e, f, g = fn(self.a_fp32, self.b_fp32, self.c_fp32, self.d_fp32)
+        self.assertEqual(e.dtype, torch.float16)
+        self.assertEqual(f.dtype, torch.float16)
+        self.assertEqual(g.dtype, torch.float16)
+
+    # TODO: fix and enable this test?
+    #   (we could technically fix this, but is it really worth it?)
+    @unittest.skipIf(True, "unsuported autocast syntax")
+    def test_reused_autocast_expr(self):
+        @torch.jit.script
+        def fn(a, b, c, d):
+            with autocast(enabled=True) as autocast_instance:
+                e = torch.mm(a, b)
+                with autocast_instance:
+                    e = torch.mm(c, d)
+                    f = torch.mm(d, e)
+            g = torch.mm(e, f)
+            return e, f, g
+        e, f, g = fn(self.a_fp32, self.b_fp32, self.c_fp32, self.d_fp32)
+        self.assertEqual(e.dtype, torch.float16)
+        self.assertEqual(f.dtype, torch.float16)
+        self.assertEqual(g.dtype, torch.float16)
+
+    def test_callees(self):
+        def helper(a, b):
```

Comment:
What if we insert a autocast scope inside the helper? Would it cause redundant cast operation here?

Reply:
Inserting an autocast scope insode helper() would be the same as having nested scopes. Currently we don't optimize casts and it's one thing I need to look into - but it's not the highest priority since no-op casts w/o copy should be idempotent and cheap, right?

Reply:
I've added a new test case to exercise this path.

In case you'd like to take a closer look at the IR before and after the autocast transformations you can:
`export PYTORCH_JIT_LOG_LEVEL=autocast`